lib/ukstore/store.c

```diff
@@ -29,3 +29,1186 @@
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_u64(const struct uk_store_entry *e, __u64 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, U64, out, u64);
+	GETCASE_UPCASTSU(e, s8, U64, out, u64);
+	GETCASE_UPCAST(e, u16, U64, out, u64);
+	GETCASE_UPCASTSU(e, s16, U64, out, u64);
+	GETCASE_UPCAST(e, u32, U64, out, u64);
+	GETCASE_UPCASTSU(e, s32, U64, out, u64);
+	GETCASE_UPCAST(e, u64, U64, out, u64);
+	GETCASE_UPCASTSU(e, s64, U64, out, u64);
+	GETCASE_UPCAST(e, uptr, U64, out, u64);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%" __SCNu64, out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_s64(const struct uk_store_entry *e, __s64 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, S64, out, s64);
+	GETCASE_UPCAST(e, s8, S64, out, s64);
+	GETCASE_UPCAST(e, u16, S64, out, s64);
+	GETCASE_UPCAST(e, s16, S64, out, s64);
+	GETCASE_UPCAST(e, u32, S64, out, s64);
+	GETCASE_UPCAST(e, s32, S64, out, s64);
+	GETCASE_DOWNCASTUS(e, u64, S64, out, s64);
+	GETCASE_UPCAST(e, s64, S64, out, s64);
+	GETCASE_DOWNCASTUS(e, uptr, S64, out, s64);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%" __SCNs64, out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_uptr(const struct uk_store_entry *e, __uptr *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s8, PTR, out, uptr);
+	GETCASE_UPCAST(e, u16, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s16, PTR, out, uptr);
+	GETCASE_UPCAST(e, u32, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s32, PTR, out, uptr);
+	GETCASE_UPCAST(e, u64, PTR, out, uptr);
+	GETCASE_UPCASTSU(e, s64, PTR, out, uptr);
+	GETCASE_UPCAST(e, uptr, PTR, out, uptr);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s16 sscanf_ret;
+
+		ret = e->get.charp(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+
+		sscanf_ret = sscanf(val, "%p", (void **) out);
+		if (sscanf_ret < 0)
+			return sscanf_ret;
+
+		return ret;
+	}
+
+	default:
+		return -EINVAL;
+	}
+}
+
+int
+_uk_store_get_charp(const struct uk_store_entry *e, char **out)
+{
+	int ret;
+	char *str;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	case uk_store_get_entry_type(u8): {
+		__u8 val;
+
+		str = calloc(UK_STORE_NR_SIZE8_LEN, sizeof(char));
+		if (unlikely(!str))
+			return -ENOMEM;
+
+		ret = e->get.u8(e->cookie, &val);
+		if (ret < 0)
+			return ret;
+		sprintf(str, "%" __PRIu8, val);
```

Comment:
if you use `snprintf` instead, the string will be `\0`-terminated in the target buffer - even if truncated but you size them big enough. You will also stay within bounds and you could just `malloc` the string buffer instead of `calloc`.